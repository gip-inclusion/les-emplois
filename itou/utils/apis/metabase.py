import time

import jwt
from django.conf import settings


ASP_SIAE_FILTER_KEY_FLAVOR1 = "identifiant_de_la_structure"
ASP_SIAE_FILTER_KEY_FLAVOR2 = "id_asp_de_la_siae"
C1_SIAE_FILTER_KEY = "identifiant_de_la_structure_(c1)"
IAE_NETWORK_FILTER_KEY = "id_r%C3%A9seau"
DEPARTMENT_FILTER_KEY = "d%C3%A9partement"
REGION_FILTER_KEY = "r%C3%A9gion"
PRESCRIBER_FILTER_KEY = "prescripteur"
JOB_APPLICATION_ORIGIN_FILTER_KEY = "origine_candidature"
PE_PRESCRIBER_FILTER_VALUE = "France Travail"
PE_FILTER_VALUE = "France Travail"

METABASE_DASHBOARDS = {
    #
    # Public stats.
    #
    "stats_public": {
        "dashboard_id": 236,
    },
    #
    # Employer stats.
    #
    "stats_siae_aci": {
        "dashboard_id": 327,
    },
    "stats_siae_etp": {
        "dashboard_id": 128,
    },
    "stats_siae_hiring": {
        "dashboard_id": 185,
        "tally_popup_form_id": "waQPkB",
        "tally_embed_form_id": "nG6J62",
    },
    "stats_siae_auto_prescription": {
        "dashboard_id": 295,
    },
    "stats_siae_follow_siae_evaluation": {
        "dashboard_id": 298,
    },
    "stats_siae_hiring_report": {
        "dashboard_id": 394,
        "tally_popup_form_id": "wb7lR1",
        "tally_embed_form_id": "wkyYjo",
    },
    #
    # Prescriber stats - CD.
    #
    "stats_cd_iae": {
        "dashboard_id": 118,
        "tally_popup_form_id": "npD4g8",
        "tally_embed_form_id": "m6ZrqY",
    },
    "stats_cd_hiring": {
        "dashboard_id": 346,
    },
    "stats_cd_brsa": {
        "dashboard_id": 330,
    },
    "stats_cd_aci": {
        "dashboard_id": 327,
    },
    #
    # Prescriber stats - PE.
    #
    "stats_pe_delay_main": {
        "dashboard_id": 168,
        "tally_popup_form_id": "3lb9XW",
        "tally_embed_form_id": "meM7DE",
    },
    "stats_pe_delay_raw": {
        "dashboard_id": 180,
    },
    "stats_pe_conversion_main": {
        "dashboard_id": 169,
        "tally_popup_form_id": "mODeK8",
        "tally_embed_form_id": "3xrPjJ",
    },
    "stats_pe_conversion_raw": {
        "dashboard_id": 182,
    },
    "stats_pe_state_main": {
        "dashboard_id": 149,
        "tally_popup_form_id": "mRG61J",
        "tally_embed_form_id": "3qLKad",
    },
    "stats_pe_state_raw": {
        "dashboard_id": 183,
    },
    "stats_pe_tension": {
        "dashboard_id": 162,
        "tally_popup_form_id": "wobaYV",
        "tally_embed_form_id": "3EKJ5q",
    },
    #
    # Institution stats - DDETS IAE - department level.
    #
    "stats_ddets_iae_auto_prescription": {
        "dashboard_id": 267,
        "tally_popup_form_id": "3qLpE2",
        "tally_embed_form_id": "nG6gBj",
    },
    "stats_ddets_iae_ph_prescription": {
        "dashboard_id": 289,
        "tally_popup_form_id": "wbWKEo",
        "tally_embed_form_id": "wvPEvQ",
    },
    "stats_ddets_iae_follow_siae_evaluation": {
        "dashboard_id": 265,
        "tally_popup_form_id": "w2XZxV",
        "tally_embed_form_id": "n9Ba6G",
    },
    "stats_ddets_iae_follow_prolongation": {
        "dashboard_id": 357,
    },
    "stats_ddets_iae_tension": {
        "dashboard_id": 389,
    },
    "stats_ddets_iae_iae": {
        "dashboard_id": 117,
    },
    "stats_ddets_iae_siae_evaluation": {
        "dashboard_id": 144,
    },
    "stats_ddets_iae_hiring": {
        "dashboard_id": 160,
        "tally_popup_form_id": "mVLBXv",
        "tally_embed_form_id": "nPpXpQ",
    },
    "stats_ddets_iae_state": {
        "dashboard_id": 310,
        "tally_popup_form_id": "w2az2j",
        "tally_embed_form_id": "3Nlvzl",
    },
    "stats_ddets_iae_aci": {
        "dashboard_id": 327,
    },
    #
    # Institution stats - DDETS LOG - department level.
    #
    "stats_ddets_log_state": {
        "dashboard_id": 310,
        "tally_popup_form_id": "w2az2j",
        "tally_embed_form_id": "3Nlvzl",
    },
    #
    # Institution stats - DREETS IAE - region level.
    #
    "stats_dreets_iae_auto_prescription": {
        "dashboard_id": 267,
        "tally_popup_form_id": "3qLpE2",
        "tally_embed_form_id": "nG6gBj",
    },
    "stats_dreets_iae_ph_prescription": {
        "dashboard_id": 289,
        "tally_popup_form_id": "wbWKEo",
        "tally_embed_form_id": "wvPEvQ",
    },
    "stats_dreets_iae_follow_siae_evaluation": {
        "dashboard_id": 265,
        "tally_popup_form_id": "w2XZxV",
        "tally_embed_form_id": "n9Ba6G",
    },
    "stats_dreets_iae_follow_prolongation": {
        "dashboard_id": 357,
    },
    "stats_dreets_iae_tension": {
        "dashboard_id": 389,
    },
    "stats_dreets_iae_iae": {
        "dashboard_id": 117,
    },
    "stats_dreets_iae_hiring": {
        "dashboard_id": 160,
        "tally_popup_form_id": "mVLBXv",
        "tally_embed_form_id": "nPpXpQ",
    },
    "stats_dreets_iae_state": {
        "dashboard_id": 310,
        "tally_popup_form_id": "w2az2j",
        "tally_embed_form_id": "3Nlvzl",
    },
    #
    # Institution stats - DRIHL - region level - IDF only.
    #
    "stats_drihl_state": {
        "dashboard_id": 310,
        "tally_popup_form_id": "w2az2j",
        "tally_embed_form_id": "3Nlvzl",
    },
    #
    # Institution stats - DGEFP - nation level.
    #
    "stats_dgefp_auto_prescription": {
        "dashboard_id": 267,
        "tally_popup_form_id": "3qLpE2",
        "tally_embed_form_id": "nG6gBj",
    },
    "stats_dgefp_follow_siae_evaluation": {
        "dashboard_id": 265,
        "tally_popup_form_id": "w2XZxV",
        "tally_embed_form_id": "n9Ba6G",
    },
    "stats_dgefp_follow_prolongation": {
        "dashboard_id": 357,
    },
    "stats_dgefp_iae": {
        "dashboard_id": 117,
    },
    "stats_dgefp_siae_evaluation": {
        "dashboard_id": 144,
    },
    "stats_dgefp_af": {
        "dashboard_id": 142,
    },
    #
    # Institution stats - DIHAL - nation level.
    #
    "stats_dihal_state": {
        "dashboard_id": 310,
        "tally_popup_form_id": "w2az2j",
        "tally_embed_form_id": "3Nlvzl",
    },
    #
    # Institution stats - IAE Network - nation level.
    #
    "stats_iae_network_hiring": {
        "dashboard_id": 301,
    },
}


# Metabase private / signed URLs
# See:
# * https://www.metabase.com/docs/latest/enterprise-guide/full-app-embedding.html
# * https://github.com/jpadilla/pyjwt
# * https://github.com/metabase/embedding-reference-apps/blob/master/django/embedded_analytics/user_stats/views.py
def _get_token(payload):
    return jwt.encode(payload, settings.METABASE_SECRET_KEY, algorithm="HS256")


def get_view_name(request):
    full_view_name = request.resolver_match.view_name  # e.g. "stats:stats_public"
    view_name = full_view_name.split(":")[-1]  # e.g. "stats_public"
    return view_name


def metabase_embedded_url(request=None, dashboard_id=None, params=None, with_title=False):
    """
    Creates an embed/signed URL for embedded Metabase dashboards:
    * expiration delay of token set at 10 minutes, kept short on purpose due to the fact that during this short time
      the user can share the iframe URL to non authorized third parties
    * do not display title of the dashboard in the iframe
    * optional parameters typically for locked filters (e.g. allow viewing data of one departement only)
    """
    if params is None:
        params = {}
    if dashboard_id is None:
        view_name = get_view_name(request)
        metabase_dashboard = METABASE_DASHBOARDS.get(view_name)
        dashboard_id = metabase_dashboard["dashboard_id"] if metabase_dashboard else None

    payload = {"resource": {"dashboard": dashboard_id}, "params": params, "exp": round(time.time()) + (10 * 60)}
    is_titled = "true" if with_title else "false"
    return settings.METABASE_SITE_URL + "/embed/dashboard/" + _get_token(payload) + f"#titled={is_titled}"
