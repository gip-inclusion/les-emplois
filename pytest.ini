[pytest]
DJANGO_SETTINGS_MODULE = config.settings.test
FAIL_INVALID_TEMPLATE_VARS = True
python_files = tests*.py test_*.py
filterwarnings =
    error
addopts =
    --reuse-db
    --strict-markers
markers =
    no_django_db: mark tests that should not be marked with django_db.
