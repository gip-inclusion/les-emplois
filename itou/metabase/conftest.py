import pytest
from django.db import connection

from itou.metabase import db


@pytest.fixture(name="metabase")
def metabase_fixture(monkeypatch):
    class FakeMetabase:
        """
        This fake metabase database allows us to benefit from all
        the Django heavy lifting that is done with creating the database,
        wrap everything in a transaction, etc.

        This makes us write the metabase tables in the main test database.

        FIXME(vperron): This is very basic for now. It still does not handle
        initial table creation, there might be table name collision or other
        issues. Let's fix them as they arise.
        """

        def __init__(self):
            self.cursor = None

        def __enter__(self):
            self.cursor = connection.cursor().cursor
            return self.cursor, connection

        def __exit__(self, exc_type, exc_value, exc_traceback):
            if self.cursor:
                self.cursor.close()

    monkeypatch.setattr(db, "MetabaseDatabaseCursor", FakeMetabase)
