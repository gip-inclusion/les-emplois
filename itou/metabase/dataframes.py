"""
Helper methods for manipulating dataframes used by both populate_metabase_emplois and
populate_metabase_fluxiae scripts.
"""
import numpy as np
import pandas as pd
from psycopg import sql
from tqdm import tqdm

from itou.metabase.db import MetabaseDatabaseCursor, create_table, get_new_table_name, rename_table_atomically


PANDA_DATAFRAME_TO_PSQL_TYPES_MAPPING = {
    np.int64: "bigint",
    np.object_: "text",
    np.float64: "double precision",
    np.bool_: "boolean",
}


def infer_columns_from_df(df):
    # Generate a dataframe with the same headers a the first non null value for each column
    df_columns = [df[column_name] for column_name in df.columns]
    non_null_values = [df_column.get(df_column.first_valid_index()) for df_column in df_columns]
    initial_line = pd.DataFrame([non_null_values], columns=df.columns)

    # Generate table sql definition from np types
    return [
        (col_name, PANDA_DATAFRAME_TO_PSQL_TYPES_MAPPING[col_type.type])
        for col_name, col_type in initial_line.dtypes.items()
    ]


def store_df(df, table_name, max_attempts=5):
    """
    Store dataframe in database.

    Do this chunk by chunk to solve
    psycopg.OperationalError "server closed the connection unexpectedly" error.

    Try up to `max_attempts` times.
    """
    # Recipe from https://stackoverflow.com/questions/44729727/pandas-slice-large-dataframe-in-chunks
    rows_per_chunk = 10 * 1000
    df_chunks = [df[i : i + rows_per_chunk] for i in range(0, df.shape[0], rows_per_chunk)]

    print(f"Storing {table_name} in {len(df_chunks)} chunks of (max) {rows_per_chunk} rows each ...")

    attempts = 0

    new_table_name = get_new_table_name(table_name)
    while attempts < max_attempts:
        try:
            columns = infer_columns_from_df(df)
            create_table(new_table_name, columns, reset=True)
            for df_chunk in tqdm(df_chunks):
                rows = df_chunk.replace({np.nan: None}).to_dict(orient="split")["data"]
                with MetabaseDatabaseCursor() as (cursor, conn):
                    with cursor.copy(
                        sql.SQL("COPY {table_name} FROM STDIN WITH (FORMAT BINARY)").format(
                            table_name=sql.Identifier(new_table_name),
                            Fields=sql.SQL(",").join(
                                [sql.Identifier(col[0]) for col in columns],
                            ),
                        )
                    ) as copy:
                        copy.set_types([col[1] for col in columns])
                        for row in rows:
                            copy.write_row(row)
                    conn.commit()
            break
        except Exception as e:
            # Catching all exceptions is a generally a code smell but we eventually reraise it so it's ok.
            attempts += 1
            print(f"Attempt #{attempts} failed with exception {repr(e)}.")
            if attempts == max_attempts:
                print("No more attemps left, giving up and raising the exception.")
                raise
            print("New attempt started...")

    rename_table_atomically(new_table_name, table_name)
    print(f"Stored {table_name} in database ({len(df)} rows).")
    print("")


def get_df_from_rows(rows):
    """
    Helper method converting rows into a dataframe.

    Rows should be a list of rows, each row being a Dict (or an OrderedDict to ensure column order) like this one:
    `{"field1": value1, "field2": value2}`
    """
    # `columns=rows[0].keys()` trick is necessary to preserve the order of columns.
    df = pd.DataFrame(rows, columns=rows[0].keys())
    return df
