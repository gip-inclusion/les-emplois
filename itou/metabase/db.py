"""
Helper methods for manipulating tables used by both populate_metabase_emplois and populate_metabase_fluxiae scripts.
"""
import logging
import os

import psycopg2
from django.conf import settings
from psycopg2 import sql
from psycopg2.extras import LoggingConnection, LoggingCursor

from itou.utils.python import timeit


logger = logging.getLogger("django.db.backends")


class MetabaseDatabaseCursor:
    def __init__(self):
        self.cursor = None
        self.connection = None

    def __enter__(self):
        self.connection = psycopg2.connect(
            host=settings.METABASE_HOST,
            port=settings.METABASE_PORT,
            dbname=settings.METABASE_DATABASE,
            user=settings.METABASE_USER,
            password=settings.METABASE_PASSWORD,
            keepalives=1,
            keepalives_idle=30,
            keepalives_interval=5,
            keepalives_count=5,
            connection_factory=LoggingConnection,
        )
        self.connection.initialize(logger)
        self.cursor = self.connection.cursor(cursor_factory=LoggingCursor if settings.SQL_DEBUG else None)
        return self.cursor, self.connection

    def __exit__(self, exc_type, exc_value, exc_traceback):
        if self.cursor:
            self.cursor.close()
        if self.connection:
            self.connection.close()


def get_current_dir():
    return os.path.dirname(os.path.realpath(__file__))


def get_new_table_name(table_name):
    return f"z_new_{table_name}"


def get_old_table_name(table_name):
    return f"z_old_{table_name}"


def rename_table_atomically(from_table_name, to_table_name):
    """
    Rename from_table_name into to_table_name.
    Most of the time, we replace an existing table, so we will first rename
    to_table_name into z_old_<to_table_name>.
    This allows to take our time filling the new table without locking the current one.
    """

    with MetabaseDatabaseCursor() as (cur, conn):
        cur.execute(
            sql.SQL("ALTER TABLE IF EXISTS {} RENAME TO {}").format(
                sql.Identifier(to_table_name),
                sql.Identifier(get_old_table_name(to_table_name)),
            )
        )
        cur.execute(
            sql.SQL("ALTER TABLE {} RENAME TO {}").format(
                sql.Identifier(from_table_name),
                sql.Identifier(to_table_name),
            )
        )
        conn.commit()
        cur.execute(sql.SQL("DROP TABLE IF EXISTS {}").format(sql.Identifier(get_old_table_name(to_table_name))))
        conn.commit()


def create_table(table_name: str, columns: list[str, str]):
    """Create table from columns names and types"""
    with MetabaseDatabaseCursor() as (cursor, conn):
        create_table_query = sql.SQL("CREATE TABLE IF NOT EXISTS {table_name} ({fields_with_type})").format(
            table_name=sql.Identifier(table_name),
            fields_with_type=sql.SQL(",").join(
                [sql.SQL(" ").join([sql.Identifier(col_name), sql.SQL(col_type)]) for col_name, col_type in columns]
            ),
        )
        cursor.execute(create_table_query)
        conn.commit()


@timeit
def build_custom_table(table_name, sql_request):
    """
    Build a new table with given sql_request.
    Minimize downtime by building a temporary table first then swap the two tables atomically.
    """
    new_table_name = get_new_table_name(table_name)
    with MetabaseDatabaseCursor() as (cur, conn):
        cur.execute(sql.SQL("DROP TABLE IF EXISTS {}").format(sql.Identifier(new_table_name)))
        conn.commit()
        cur.execute(sql.SQL("CREATE TABLE {} AS {}").format(sql.Identifier(new_table_name), sql.SQL(sql_request)))
        conn.commit()

    rename_table_atomically(new_table_name, table_name)


def build_final_tables():
    """
    Build final custom tables one by one by playing SQL requests in `sql` folder.

    Typically:
    - 001_fluxIAE_DateDerniereMiseAJour.sql
    - 002_missions_ai_ehpad.sql
    - ...

    The numerical prefixes ensure the order of execution is deterministic.

    The name of the table being created with the query is derived from the filename,
    # e.g. '002_missions_ai_ehpad.sql' => 'missions_ai_ehpad'
    """
    path = f"{get_current_dir()}/sql"
    for filename in sorted([f for f in os.listdir(path) if f.endswith(".sql")]):
        print(f"Running {filename} ...")
        table_name = "_".join(filename.split(".")[0].split("_")[1:])
        with open(os.path.join(path, filename), "r") as file:
            sql_request = file.read()
        build_custom_table(table_name=table_name, sql_request=sql_request)
        print("Done.")
