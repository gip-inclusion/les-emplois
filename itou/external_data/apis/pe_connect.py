import logging

import requests
from django.conf import settings
from django.db import transaction
from django.forms.models import model_to_dict
from django.utils.dateparse import parse_datetime

from itou.external_data.models import ExternalDataImport, JobSeekerExternalData


# PE Connect API data retrieval tools

API_ESD_BASE_URL = settings.API_ESD_BASE_URL

ESD_USERINFO_API = "peconnect-individu/v1/userinfo"
ESD_COORDS_API = "peconnect-coordonnees/v1/coordonnees"
ESD_STATUS_API = "peconnect-statut/v1/statut"
ESD_BIRTHDATE_API = "peconnect-datenaissance/v1/etat-civil"

# FIXME: not registered yet
ESD_COMPENSATION_API = "peconnect-indemnisations/v1/indemnisation"
ESD_PT_TRAININGS_API = "peconnect-formations/v1/formations"
ESD_PT_LICENSES_API = "peconnect-formations/v1/permits"

# Internal
logger = logging.getLogger(__name__)


# This part may be refactored with the processing of other APIs
# YAGNI at the moment


def _call_api(api_path, token):
    """
    Make a sync call to an API
    For further processing, returning something else than `None` is considered a success
    """
    url = f"{API_ESD_BASE_URL}/{api_path}"
    response = requests.get(url, headers={"Authorization": f"Bearer {token}"}, timeout=settings.REQUESTS_TIMEOUT)
    if response.status_code == 200:
        result = response.json()
        # logger.debug(f"CALL {url}: {result}")
        return result

    # Track it for QoS
    logger.warning("API call to: %s returned status code %s", url, response.status_code)
    return None


def _fields_or_failed(result, keys):
    if not result:
        return None

    return {k: v for k, v in result.items() if k in keys}


def _get_userinfo(token):
    """
    Get main info from user:
        * first and family names
        * gender
        * email address

    See: https://www.emploi-store-dev.fr/portail-developpeur-cms/home/catalogue-des-api/documentation-des-api/api/api-pole-emploi-connect/api-peconnect-individu-v1.html  # noqa E501
    """
    # Fields of interest
    keys = ["given_name", "family_name", "gender", "email"]
    return _fields_or_failed(_call_api(ESD_USERINFO_API, token), keys)


def _get_birthdate(token):
    """
    Get birthdate of user (format `YYYY-MM-DDTHH:MM:SSZ`), converted as `datetime` object.

    See: https://www.emploi-store-dev.fr/portail-developpeur-cms/home/catalogue-des-api/documentation-des-api/api/api-pole-emploi-connect/api-peconnect-datenaissance-v1.html  # noqa E501
    """
    key = "dateDeNaissance"
    # code, resp = _call_api(ESD_BIRTHDATE_API, token)
    result = _fields_or_failed(_call_api(ESD_BIRTHDATE_API, token), [key])
    if result:
        return {key: result.get(key)}

    return None


def _get_status(token):
    """
    Get current status of candidate.

    Returns a dict with codeStatutIndividu field from API:
        * 0: does not seek a job
        * 1: active jobseeker

    See: https://www.emploi-store-dev.fr/portail-developpeur-cms/home/catalogue-des-api/documentation-des-api/api/api-pole-emploi-connect/api-peconnect-statut-v1.html  # noqa E501
    """
    key = "codeStatutIndividu"
    result = _fields_or_failed(_call_api(ESD_STATUS_API, token), [key])
    if result:
        code = result.get(key)
        return {key: int(code)}

    return None


def _get_address(token):
    """
    Get current address of the candidate:

    Returns a dict with fields:
        * adresse1
        * adresse2
        * adresse3
        * adresse4
        * codePostal
        * codeINSEE
        * libelleCommune

    Does not return country related fields (only France ATM)

    See: https://www.emploi-store-dev.fr/portail-developpeur-cms/home/catalogue-des-api/documentation-des-api/api/api-pole-emploi-connect/api-peconnect-coordonnees-v1.html   # noqa E501
    """
    keys = ["adresse1", "adresse2", "adresse3", "adresse4", "codePostal", "codeINSEE", "libelleCommune"]
    return _fields_or_failed(_call_api(ESD_COORDS_API, token), keys)


def _get_compensations(token):
    """
    Get user "compensations" (social allowance):

    Return a dict with fields:
        * beneficiairePrestationSolidarite (has one or more of AER, AAH, ASS, RSA)
        * beneficiaireAssuranceChomage (has ARE or ASP)

    See: https://www.emploi-store-dev.fr/portail-developpeur-cms/home/catalogue-des-api/documentation-des-api/api/api-pole-emploi-connect/api-indemnisations-v1.html  # noqa E501
    """
    keys = ["beneficiairePrestationSolidarite", "beneficiaireAssuranceChomage"]
    return _fields_or_failed(_call_api(ESD_COMPENSATION_API, token), keys)


def get_aggregated_user_data(token):
    """
    Aggregates all needed user data before formatting and storage.
    Returns a pair status and a "flat" dict.
    """
    # Include API results "à volonté"
    results = [
        _get_userinfo(token),
        _get_birthdate(token),
        _get_status(token),
        _get_address(token),
        _get_compensations(token),
    ]

    ok = all(results)
    partial = not ok and any(results)
    cleaned_results = [part for part in results if part]

    if ok:
        status = ExternalDataImport.STATUS_OK
    elif partial:
        status = ExternalDataImport.STATUS_PARTIAL
    else:
        status = ExternalDataImport.STATUS_FAILED

    user_data = {}
    for result in cleaned_results:
        user_data.update(result)

    return status, user_data


def _model_fields_changed(initial, final_instance):
    """
    Return an array with fields of the model_instance object that will be changed on `save()`
    """
    final = model_to_dict(final_instance)
    class_name = type(final_instance).__name__
    return [f"{class_name}/{final_instance.pk}/{k}" for k, v in initial.items() if v != final.get(k)]


# External user data from PE Connect API:
# * transform raw data from API
# * dispatch data into models if possible
# * or store as key / value if needed


def set_pe_data_import_from_user_data(pe_data_import, user, status, user_data):
    """
    Store user data and produce a "report" containing a JSON object, with these fields:
        - fields_fetched (array): successfully imported field names
        - fields_failed (array): fields that could not be fetched (API error...)
        - fields_updated (array): updated fields in the db each of form:
            `class_name/pk/field_name` (f.i. `User/10/birthdate`)

    Return a ExternalDataImport object containing outcome of the data import
    """
    fields_fetched = [k for k, v in user_data.items() if v is not None]
    fields_failed = [k for k, v in user_data.items() if v is None]

    job_seeker_data, _ = JobSeekerExternalData.objects.get_or_create(user=user, data_import=pe_data_import)

    # Record changes on model objects:
    initial_job_seekeer_data = model_to_dict(job_seeker_data)
    initial_user = model_to_dict(user)

    for k in fields_fetched:
        v = user_data.get(k)

        # User part:
        if k == "dateDeNaissance":
            new_value = user.birthdate or parse_datetime(v)
            user.birthdate = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "birthdate", new_value)
        elif k == "adresse4":
            new_value = "" or user.address_line_1 or v
            user.address_line_1 = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "address_line_1", new_value)
        elif k == "adresse2":
            new_value = "" or user.address_line_2 or v
            user.address_line_2 = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "address_line_2", new_value)
        elif k == "codePostal":
            new_value = user.post_code or v
            user.post_code = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "post_code", new_value)
        elif k == "libelleCommune":
            new_value = user.city or v
            user.city = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "city", new_value)

        # JobSeekerExternalData part:
        if k == "codeStatutIndividu":
            new_value = v == 1
            job_seeker_data.is_pe_jobseeker = new_value
            user.update_external_data_source_history_field(settings.PROVIDER_PE_CONNECT, "is_pe_jobseeker", new_value)
        elif k == "beneficiairePrestationSolidarite":
            job_seeker_data.has_minimal_social_allowance = v
            user.update_external_data_source_history_field(
                settings.PROVIDER_PE_CONNECT, "has_minimal_social_allowance", v
            )

    # Check updated fields
    # To be done before saving objects:
    fields_updated = _model_fields_changed(initial_user, user) + _model_fields_changed(
        initial_job_seekeer_data, job_seeker_data
    )

    # Atomicity in outer call
    user.save()
    job_seeker_data.save()

    report = {"fields_fetched": fields_fetched, "fields_failed": fields_failed, "fields_updated": fields_updated}
    pe_data_import.status = status
    pe_data_import.report = report
    return pe_data_import


#  Public


def import_user_pe_data(
    user,
    token,
    pe_data_import=None,
):
    """
    Import external user data via PE Connect
    Returns a valid ExternalDataImport object when result is PARTIAL or OK.
    """
    if not pe_data_import:
        pe_data_import = ExternalDataImport.objects.create(
            source=ExternalDataImport.DATA_SOURCE_PE_CONNECT, user=user, status=ExternalDataImport.STATUS_PENDING
        )

        try:
            # External requests
            status, user_data = get_aggregated_user_data(token)
            with transaction.atomic():
                # A refactoring would be helpful to split user/job seeker saving
                # and to use the status before calling save()
                # Save user and job seeker data
                set_pe_data_import_from_user_data(pe_data_import, user, status, user_data)
                pe_data_import.save()

            if status == ExternalDataImport.STATUS_OK:
                logger.info("Stored external data for user %s", user)
            elif status == ExternalDataImport.STATUS_PARTIAL:
                logger.warning("Could only fetch partial results for %s", user)
            else:
                logger.error("Could not fetch any data for %s: not data stored", user)
        except Exception as e:
            logger.error("Data import for %s failed: %s", user, e)
            pe_data_import.status = ExternalDataImport.STATUS_FAILED
            pe_data_import.save()

    return pe_data_import
