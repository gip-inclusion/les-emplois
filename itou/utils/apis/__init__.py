from django.conf import settings

from itou.utils.apis.pole_emploi import PoleEmploiRoyaumePartenaireApiClient


def pole_emploi_api_client():
    return PoleEmploiRoyaumePartenaireApiClient(
        settings.API_ESD["BASE_URL"],
        settings.API_ESD["AUTH_BASE_URL"],
        settings.API_ESD["KEY"],
        settings.API_ESD["SECRET"],
    )
