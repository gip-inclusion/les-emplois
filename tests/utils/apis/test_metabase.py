import pytest

from itou.utils.apis.metabase import Client


@pytest.fixture(name="client")
def client_fixture(settings):
    settings.METABASE_API_KEY = "metabase-api-key"
    return Client("http://metabase")


@pytest.fixture(name="data_results")
def data_results_fixture():
    return [
        {"Col 1": "Value 1", "Col 2": 1, "Col 3": 1.0},
        {"Col 1": "Value 2", "Col 2": 2, "Col 3": 2.0},
    ]


@pytest.fixture(name="card_results")
def card_results_fixture():
    return {
        "dataset_query": {
            "query": {
                "filter": [],
                "breakout": [],
            }
        }
    }


@pytest.mark.parametrize("url", ["http://metabase", "http://metabase/"])
def test_client_init(faker, settings, url):
    settings.METABASE_API_KEY = faker.word()

    client = Client(url)
    assert client._client.base_url == "http://metabase/api/"
    assert client._client.headers["X-API-KEY"] == settings.METABASE_API_KEY
    assert client._client.timeout.as_dict() == {
        "connect": 5,
        "read": 60,
        "write": 5,
        "pool": 5,
    }


@pytest.mark.respx(base_url="http://metabase/api")
def test_fetch_card_results_without_filters_nor_group_by(respx_mock, client, data_results):
    respx_mock.post("/card/42/query/json").respond(202, json=data_results)
    assert client.fetch_card_results(42) == data_results


@pytest.mark.respx(base_url="http://metabase/api")
def test_fetch_card_results_with_filters(snapshot, respx_mock, client, data_results):
    card_api_route = respx_mock.get("/card/42")
    dataset_api_route = respx_mock.post("/dataset/json").respond(202, json=data_results)

    # Without pre-existing filters
    card_api_route.respond(200, json={"dataset_query": {"query": {"foo": "bar"}}})
    client.fetch_card_results(42, filters={21: ["baz"]})
    assert dataset_api_route.calls.last.request.content == snapshot(name="without pre-existing")

    # With pre-existing filters
    card_api_route.respond(
        200,
        json={
            "dataset_query": {
                "query": {
                    "filter": ["and", ["=", Client._build_metabase_field(100), "filter value 1", "filter value 2"]],
                }
            }
        },
    )
    client.fetch_card_results(42, filters={21: ["baz"]})
    assert dataset_api_route.calls.last.request.content == snapshot(name="with pre-existing")


@pytest.mark.respx(base_url="http://metabase/api")
def test_fetch_card_results_with_group_by(snapshot, respx_mock, client, data_results):
    card_api_route = respx_mock.get("/card/42")
    dataset_api_route = respx_mock.post("/dataset/json").respond(202, json=data_results)

    # Without pre-existing breakout
    card_api_route.respond(200, json={"dataset_query": {"query": {"foo": "bar"}}})
    client.fetch_card_results(42, group_by=[1])
    assert dataset_api_route.calls.last.request.content == snapshot(name="without pre-existing")

    # With pre-existing breakout
    card_api_route.respond(
        200,
        json={
            "dataset_query": {
                "query": {
                    "breakout": [client._build_metabase_field(100)],
                }
            }
        },
    )
    client.fetch_card_results(42, group_by=[2])
    assert dataset_api_route.calls.last.request.content == snapshot(name="with pre-existing")
