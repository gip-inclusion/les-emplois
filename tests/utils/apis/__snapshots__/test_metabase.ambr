# serializer version: 1
# name: test_fetch_card_results_with_filters[with pre-existing]
  b'query=%7B%22query%22%3A+%7B%22filter%22%3A+%5B%22and%22%2C+%5B%22and%22%2C+%5B%22%3D%22%2C+%5B%22field%22%2C+100%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%2C+%22filter+value+1%22%2C+%22filter+value+2%22%5D%5D%2C+%5B%5B%22%3D%22%2C+%5B%22field%22%2C+21%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%2C+%22baz%22%5D%5D%5D%7D%7D'
# ---
# name: test_fetch_card_results_with_filters[without pre-existing]
  b'query=%7B%22query%22%3A+%7B%22foo%22%3A+%22bar%22%2C+%22filter%22%3A+%5B%22and%22%2C+%5B%5D%2C+%5B%5B%22%3D%22%2C+%5B%22field%22%2C+21%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%2C+%22baz%22%5D%5D%5D%7D%7D'
# ---
# name: test_fetch_card_results_with_group_by[with pre-existing]
  b'query=%7B%22query%22%3A+%7B%22breakout%22%3A+%5B%5B%22field%22%2C+100%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%2C+%5B%22field%22%2C+2%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%5D%7D%7D'
# ---
# name: test_fetch_card_results_with_group_by[without pre-existing]
  b'query=%7B%22query%22%3A+%7B%22foo%22%3A+%22bar%22%2C+%22breakout%22%3A+%5B%5B%22field%22%2C+1%2C+%7B%22base-type%22%3A+%22type%2FText%22%7D%5D%5D%7D%7D'
# ---
